### DISCUSSION TITLE
Give us the ability to copy or reuse blocks

### DISCUSSION THREAD

**Discussion by prjseal** [umbraco/Umbraco-CMS#11750] [Features and ideas]
I have been creating a lot of blocks lately using the block list data type. I think it would be really useful to be able to copy or reuse blocks.

Here is an example. 

Let's say I have a block called Bulleted List Content and I want to reuse that block on another block list.

![image](https://user-images.githubusercontent.com/9142936/145429242-67296258-1dab-40c4-b701-c7b34c9f4d0f.png)

At the moment I would have to create that block again manually on the other block list data type.

It would be really good if we could do 1 or both of these ideas.

## Idea 1
Allow us to copy a block configuration a bit like how you can copy a block's contents. I imagine a copy icon next to the delete icon

![image](https://user-images.githubusercontent.com/9142936/145429530-898fdb09-9c51-4ca0-aa93-ff7bf6cd1173.png)

Then we go to our new block list and when we click on add we get an option to add from clipboard, again a bit like the copy and paste ability we've got for block content.

## Idea 2
Give us the ability to create blocks definitions in a central place that we can then pick from to add to our block lists. 

I would love to hear your thoughts about this one and see if we can get one or both of these ideas implemented in the core.

**Comment by AndyBoot** [comment]
Huge +1 for Idea 1. Having used other CMS's and page builders this is a fairly standard and highly used feature which is much used by content editors. I'd love to see this implemented at some point in the near future.

**Comment by skttl** [comment]
+1 for idea 2. Would make it a lot easier to tailor block list setups for different document types.

**Comment by ault** [comment]
Both are appealing. Though the second might be the more satisfying outcome, the first (simple copy) option seems to match the existing Umbraco UIX. (i.e. you can right-click/copy doctypes, why not datatypes?)

**Comment by nathanwoulfe** [comment]
👋👋 I've started working on a package to solve (partially at least) the second point - being able to define a block once and reuse it on multiple content nodes.

This was part of the original Block List concept from the 2019 retreat, where we envisioned a separate tree specifically for hosting content elements. I don't know if this is on HQ's roadmap, and I needed the functionality for a project, so figured I'd make it public.

Haven't released anything as yet, needs a little more polish, but it's largely working.

Code is at github.com/nathanwoulfe/blip

**Reply by jesperordrup** [reply]
@nathanwoulfe is blip private repo?

**Reply by nathanwoulfe** [reply]
Oh, possibly... Will fix!

**Reply by deanleigh** [reply]
So basically Block Level Content Templates?
I would like these.

**Comment by jesperordrup** [comment]
> I have been creating a lot of blocks lately using the block list data type. I think it would be really useful to be able to copy or reuse blocks.
> 
> Here is an example.
> 
> Let's say I have a block called Bulleted List Content and I want to reuse that block on another block list.
> 
> ![image](https://user-images.githubusercontent.com/9142936/145429242-67296258-1dab-40c4-b701-c7b34c9f4d0f.png)
> 
> At the moment I would have to create that block again manually on the other block list data type.
> 
> It would be really good if we could do 1 or both of these ideas.
> 
> ## Idea 1
> Allow us to copy a block configuration a bit like how you can copy a block's contents. I imagine a copy icon next to the delete icon
> 
> ![image](https://user-images.githubusercontent.com/9142936/145429530-898fdb09-9c51-4ca0-aa93-ff7bf6cd1173.png)
> 
> Then we go to our new block list and when we click on add we get an option to add from clipboard, again a bit like the copy and paste ability we've got for block content.
> 
> ## Idea 2
> Give us the ability to create blocks definitions in a central place that we can then pick from to add to our block lists.
> 
> I would love to hear your thoughts about this one and see if we can get one or both of these ideas implemented in the core.

Idea 1: copy blocks? You can already afaik. 

Love idea 2.

**Reply by prjseal** [reply]
You can only copy block content, I want to copy the configuration, so I don't have to create the same block on a different block list

**Reply by jesperordrup** [reply]
Ahhh my bad - now I get it. True - that would be nice if it was supported by UI. But you can do it today. Just install usync :-) and edit the files. It's much easier.

Im gonna beat the dead 💀 horse 🐴   and say again that all Umbraco configuration especially doctypes and properties ideally should be stored in config files, not in db. 😆  But luckily we have uSync 🥇

**Comment by kmeilander** [comment]
I've been thinking about something like this too, although I think I'm using the block editor a little differently than other people so I would be worried that option 2 would actually make more work for me and I would more lean towards copying and pasting option. 

Here's my setup.   I don't use Settings within my block view at all, and instead make sure that every thing I need for that is in my Content doctype - even if it's something that changes the UI.   (I've never understood why you would want a 2 data models instead of 1 to render out your block)

Instead, my Settings are for setting _how_ the block is added to the page in the blocklist view renderer.  For each blocklist view I create a Settings doctype, and then use that Setting on all blocks within the BlockList that will use that render view (ie. Blocklist - Row, Blocklist - Column).  Then within the blocklist view, I look at the Settings for each block, and write out code for the generic placeholder settings such as spacing, wrap in container, background styles, boarders, css classes, or html element names.  And then render the Block Partial.

(I set the template in my GetBlockListHtml method)
`@Html.GetBlockListHtml(Model.Components, "rows")`

(And my  Blocklist view looks something like this)
```
<div class="umb-block-list umb-block-list--rows">
    @foreach (var block in Model)
    {
        if (block?.ContentUdi == null) { continue; }

        var data = block.Content;
        var settings = block.Settings as ContentModels.BlockSettingsRow;

        var paddingTop = settings != null ? string.Format(paddingMap[settings.PaddingTop], "padding-top") : "";
        var paddingBottom = settings != null ? string.Format(paddingMap[settings.PaddingBottom], "padding-bottom") : "";

        <div class="@settings?.Class @paddingBottom @paddingTop" data-blocktype="@data.ContentType.Alias">
            @await Html.PartialAsync("BlockList/Components/" + data.ContentType.Alias, block)
        </div>
            
    }
</div>

```



Is anyone else here using a similar setup for their blocklist? 

I'm assuming that the central place for option 2 would also include Settings on the blocks, which means I would have to make a separate block for each use case so really nothing will actually be reused for me (unless I can convince everyone not to use the Settings Model as part of the block).  

I know copying and pasting isn't idea since it loses the relationship - so if you have already made an copy, if you update like the label on one block, it wouldn't update on the copy.  But for me, that would be needed so I can update the block settings to be what I need.

**Comment by bjarnef** [comment]
Previous I worked on the ability to copy block configuration, but didn't got much feedback from HQ and closed it now new features are focuses on v9:
https://github.com/umbraco/Umbraco-CMS/pull/10565

**Reply by prjseal** [reply]
That’s so cool. I hope they merge it this time. Thanks for making it already

**Reply by bjarnef** [reply]
Yeah, not finished though but more a POC. I think we need some feedback from HQ how the block configuration should be copied, I imagine it would be a separate `TYPES.BLOCK_CONFIG`:
https://github.com/umbraco/Umbraco-CMS/blob/v9/contrib/src/Umbraco.Web.UI.Client/src/common/services/clipboard.service.js#L17-L20

Furthermore should pasting be part of the "Add" dialog or a specific button next to the "Add" button.

A custom action to copy all block configuration items as property actions would also be useful to copy most block configurations to another block list instance - especially when copying a datatype instance isn't a feature yet.
